src/state/opamGlobalState.ml

```diff
@@ -154,6 +160,9 @@ let drop gt =
   let _ = unlock gt in ()
 
 let with_write_lock ?dontblock gt f =
+  if OpamStateConfig.is_readonly_opamroot gt then
+    OpamConsole.error_and_exit `Configuration_error
+      "Global state try to get write loçck whil opam root is readonly";
```

Comment:
```suggestion
      "The opam root has been upgraded by a newer version of opam-state and cannot be written to";
```

Reply:
I would like to be heading towards a world where only opam-client uses `OpamConsole` - can we turn this into an exception which is caught in the client instead? (thinking of users of opam-state itself)

Reply:
After discussion, moving all exceptions in opam-client is post poned. For plugins, opam exception can be catched.